vignettes/sir_models.Rmd

```diff
@@ -0,0 +1,680 @@
+- A model to fit, which must be a dust generator.
+- A comparison function, which is an R function which calculates the likelihood of the state given the data.
+
+## Model data
+
+To demonstrate this, we will attempt to infer the parameters $\beta$ and $\gamma$ using daily case counts. In reality these counts would be measured, but here we generate them from a single run of the model, and adding some constant noise proportional to the epidemic size:
+
+```{r}
+# Set up model
+dt <- 0.25
+inv_dt <- 1/dt
+sir <- gen_sir$new(data = list(dt = dt, S_ini = 1000, I_ini = 10, beta = 0.2, gamma = 0.1),
+                   step = 0,
+                   n_particles = 1L,
+                   n_threads = 1L,
+                   seed = 2L)
+
+# Run model forward 100 days, using 4 steps a day
+# Calculate the number of new cases by counting how many susceptibles have been depleted between updates
+n_steps <- 100
+incidence <- rep(NA, length.out = n_steps)
+true_history <- array(NA_real_, c(sir$info()$len, 1, n_steps + 1))
+true_history[, 1, 1] <- sir$state()
+for (t in seq_len(n_steps)) {
+    state_start <- sir$state()
+    state_end <- sir$run(t * inv_dt)
+    true_history[, 1, t + 1] <- state_end
+    incidence[t] <- state_start[2, 1] - state_end[2, 1]
+}
+
+# Add very simple form of noise
+incidence <- incidence + 
+  round(rnorm(mean = 0, n = length(true_history[3,1,-1]), sd = 0.1*sqrt(true_history[3,1,-1])))
+incidence <- unlist(lapply(incidence, max, 0))
+incidence_data <- data.frame(cases = incidence, day = seq_len(n_steps))
+incidence
+```
+
+This can then be processed with `particle_filter_data()` as follows:
+
+```{r paged.print=TRUE, layout="l-body-outset"}
+library(mcstate)
+sir_data <- particle_filter_data(data = incidence_data,
+                                 time = "day", 
+                                 rate = inv_dt)
+paged_table(sir_data)
+```
+
+More complex models may have additional data streams added as further columns (e.g. deaths, hospital admissions). This will be fed as observed data to the compare function, which will combine and compute the evidence for the current model state.
+
+## Defining the comparison function
+
+The comparison function will typically be a likelihood, the probability of the simulation run given the data. This can be any function in R which takes the current simulated `state`, the state at the previous time step `prev_state`, a `data` row for the current time point and a list of any parameters `pars` the function needs.
+
+For this model the number of cases can be calculated from the change in susceptibles between time steps (as above). In this comparison function, the number of cases within each time interval is modelled as being Poisson distributed. A small amount of noise is added to prevent zero expectations.
+
+```{r}
+case_compare <- function(state, prev_state, observed, pars = NULL) {
+  if (is.na(observed$cases)) {
+    return(NULL)
+  }
+  if (is.null(pars$compare$exp_noise)) {
```

Comment:
This if / else is complicating things here and was needed I think in order to help with testing. If you want to demonstrate tuning the parameter I would unconditionally use your else branch, otherwise I'd unconditionally use your if branch